src/openzeppelin/utils/structs/checkpoints.cairo

```diff
@@ -0,0 +1,213 @@
+use integer::u32_sqrt;
+use openzeppelin::utils::math;
+use super::vec::VecTrait;
+use super::vec::Vec;
+use traits::Into;
+
+/// `Trace` struct, for checkpointing values as they change at different points in
+/// time, and later looking up past values by block number. See {Votes} as an example.
+struct Trace {
+    _checkpoints: Vec<Checkpoint>
+}
+
+/// Generic checkpoint representation.
+#[derive(Drop, Copy)]
+struct Checkpoint {
+    _key: u64,
+    _value: u256
+}
+
+#[generate_trait]
+impl TraceImpl of TraceTrait {
+    /// Pushes a (`key`, `value`) pair into a Trace so that it is stored as the checkpoint.
+    ///
+    /// Returns previous value and new value.
+    fn push(ref self: Trace, key: u64, value: u256) -> (u256, u256) {
+        self._checkpoints._insert(key, value)
+    }
+
+    /// Returns the value in the first (oldest) checkpoint with key greater or equal
+    /// than the search key, or zero if there is none.
+    fn lower_lookup(self: Trace, key: u64) -> u256 {
```

Comment:
For Votes you will only need upper lookup. For focus I would suggest removing the lower lookups.